brave/brave-core
# Refactor DnD on NTP Top Sites

Close brave/brave-browser#2971

## Submitter Checklist:

- Verified that these changes build without errors on
- Verified that these changes pass automated tests (unit, browser, security tests) on

## Test Plan:

### Assertions:

- Fresh profile should have 0 entries
- Non-fresh profile should show your current top sites unpinned
- Should be able to drag n' drop tiles without errors
- Should not be able to drag a pinned tile
- Should not be able to drop over a pinned tile
- Changes should persist on page refresh
- Should be able to exclude a site
- Should be able to undo an excluded site, which should go back to the first grid position
- Should be able to undo all excluded sites, which should go back to first grid positions
- Should be able to spot a bookmarked site on first render
- Should be able to toggle a bookmarked site

### Manual test plan:

**Note**: Chromium topSites API is not reliable and can drastically vary the time to update the list. Below steps will be async so it's not guaranteed that you will se the site after you visit it, but they should work after some site interaction. 

1. Fresh profile
2. Visit https://brave.com
3. Visit https://quora.com
4. Once sites are available on the grid, pin the site on the 2nd grid position
5. Visit https://twitter.com
6. Pinned site should retain position: grid should be Twitter (1), `PINNED_SITE` (2), other site (3)

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.
